### ISSUE TITLE
connectiontimeout  error can not be recovery even the server is ok and Resolver refresh address

### ISSUE THREAD

**Issue by fengyeall111** [grpc/grpc-dotnet#2125]
### What version of gRPC and what language are you using?
2.51.0
### What operating system (Linux, Windows,...) and version?

windows
### What runtime / compiler are you using (e.g. .NET Core SDK version `dotnet --info`)
.net 7.0

### What did you do?
```
1.  var  channel = GrpcChannel.ForAddress(address, new GrpcChannelOptions
                            {
                                HttpHandler = new SocketsHttpHandler
                                {
                                    EnableMultipleHttp2Connections = true,
                                    PooledConnectionIdleTimeout = TimeSpan.FromSeconds(60),
                                    ConnectTimeout = TimeSpan.FromSeconds(5),  
                                },
                                Credentials = credentials,  
                                ServiceConfig = new ServiceConfig
                                {
                                    LoadBalancingConfigs = {
                                        new BalanceConfig(opt.BalancerName)
                                    }
                                },
                                LoggerFactory = _loggerFactory,
                                ServiceProvider = services.BuildServiceProvider(),
                            });
//  dns   Resolver implements like this:
 public override void Start(Action<ResolverResult> listener)
        {
            _listener = listener
        }

 public override void Refresh()
        {      
             var balanceAddr = new BalancerAddress(temp[0], port > 0 ? port: 5002 );
             var list = new List<BalancerAddress>();
               _listener(list.AsReadOnly()); 
        }
//  load balance implements SubchannelLoadBalancer class
```
2.  use factory and channel create client
```
 CallInvoker callInvoker = channel.CreateCallInvoker().
  instance = GrpcClientFactory.CreateGrpcService<T>(callInvoker);
```
3. use client to send rpc request, without running server,error like this:
```
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|281_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
   at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)")
   at Grpc.Net.Client.Balancer.Internal.ConnectionManager.PickAsync(PickContext context, Boolean waitForReady, CancellationToken cancellationToken)
   at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
   at TL.Grpc.GrpcClientInterceptor.HandleResponse[TResponse](Task`1 t, ISpan span) in C:\workspace\csharp\TL.Grpc\Source\GrpcClientInterceptor.cs:line 49
   at ProtoBuf.Grpc.Internal.Reshape.UnaryTaskAsyncImpl[TRequest,TResponse](AsyncUnaryCall`1 call, MetadataContext metadata, CancellationToken cancellationToken) in /_/src/protobuf-net.Grpc/Internal/Reshape.cs:line 549
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at LBGrpcClient.Program.RunStressClient(Boolean isBalance) in C:\workspace\csharp\TL.Grpc\LBGrpcClient\Program.cs:line 50
```
4. start grpc server listen on server

5.  wait some second，send rpc request，but the client also con not connect to grpc server ，produce error like 
```
 Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.TimeoutException: A connection could not be established within the configured ConnectTimeout.
       ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
         at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)
         --- End of inner exception stack trace ---")
         at Grpc.Net.Client.Balancer.Internal.ConnectionManager.PickAsync(PickContext context, Boolean waitForReady, CancellationToken cancellationToken)
         at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
info: Grpc.Net.Client.Internal.GrpcCall[3]
      Call failed with gRPC error status. Status code: 'Unavailable', Message: 'Error connecting to subchannel.'.
System.AggregateException: One or more errors occurred. (Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.TimeoutException: A connection could not be established within the configured ConnectTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)
   --- End of inner exception stack trace ---"))
 ---> Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.TimeoutException: A connection could not be established within the configured ConnectTimeout.
 ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)
   --- End of inner exception stack trace ---")
   at Grpc.Net.Client.Balancer.Internal.ConnectionManager.PickAsync(PickContext context, Boolean waitForReady, CancellationToken cancellationToken)
   at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
   at TL.Grpc.GrpcClientInterceptor.HandleResponse[TResponse](Task`1 t, ISpan span) in C:\workspace\csharp\TL.Grpc\Source\GrpcClientInterceptor.cs:line 49
   at ProtoBuf.Grpc.Internal.Reshape.UnaryTaskAsyncImpl[TRequest,TResponse](AsyncUnaryCall`1 call, MetadataContext metadata, CancellationToken cancellationToken) in /_/src/protobuf-net.Grpc/Internal/Reshape.cs:line 560
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait(CancellationToken cancellationToken)
   at LBGrpcClient.Program.RunStressClient(Boolean isBalance) in C:\workspace\csharp\TL.Grpc\LBGrpcClient\Program.cs:line 50
fail: Grpc.Net.Client.Internal.GrpcCall[6]
      Error starting gRPC call.
      Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.TimeoutException: A connection could not be established within the configured ConnectTimeout.
       ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
         at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)
         --- End of inner exception stack trace ---")
         at Grpc.Net.Client.Balancer.Internal.ConnectionManager.PickAsync(PickContext context, Boolean waitForReady, CancellationToken cancellationToken)
         at Grpc.Net.Client.Balancer.Internal.BalancerHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
         at Grpc.Net.Client.Internal.GrpcCall`2.RunCall(HttpRequestMessage request, Nullable`1 timeout)
```
### Anything else we should know about your project / environment?
I have debug the source code 'grpc.net.client', and the subchannel actually receive address update ,but  socketconnectivitySubchannelTransport  always returns timeout error ,even  address  refresh.
So, I guess there is some bug exist in the subchannelloadbalancer.but I can not find exactly errror
1.dns reslover return addr "127.0.0.1:5002"
2. server not stop
3.  client send rpc request ,of course, connectimeout error will be reported 
4.  start server
5. send rpc request
6. the result is also  connectimeout error
```

**Comment by JamesNK** [comment]
I think this bug was fixed in 2.52.0 - https://github.com/grpc/grpc-dotnet/pull/1998

Update to the latest version and retry.

**Comment by fengyeall111** [comment]
you are right，thank you very much.
